enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,214 @@
+differently between OVN-Kubernetes and other CNIs.
+
+## Design Details
+
+The Service of loadblancer type is designed to exposes the Service externally
+using a cloud provider's load balancer. Normally, there is a loadbalancer
+Service controller which is responsible for:
+
+1. provision a loadbalancer for the Service.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+However, in the context of MicroShift, as we only get one node in the cluster,
+it doesn't make sense to use a real load balancer to forward the ingress
+traffic. So instead of provisioning the loadbalancer instance, the controller
+is only responsible for helping the CNIs to plumbing the ingress traffic.
+
+### Traffic Plumbing
+
+Since different CNIs get different implementation on how to support Kubernetes
+Services. We need to do the traffic plumbing differently for different CNI. The
+controller shall be aware of the CNI type from the configuration and behave
+accordingly.
+
+#### OVN-Kubernetes
+
+OVN-Kubernetes has already implemented the [loadbalancer ingress
+IP](https://github.com/openshift/ovn-kubernetes/blob/master/docs/external-ip-and-loadbalancer-ingress.md)
+support. The ovnkube-node component which is running on each host watches the
+Service's `.status.loadBalancer` field, and insert iptables rules that does DNAT
+to the clusterIP of the Service. Therefore, from the loadbalancer Service
+controller perspective, it will only be responsible for the updating the status
+of the Service. The ovn-kubernetes will do the rest of the job.
+
+#### Other CNIs
+
+For other CNIs, which don't have the similar function as OVN-Kubernetes, neither
+do they bypass the kernel for the ingress traffic, we can take the similar
+approach as K3S. The K3S's loadbalancer Service controller provisions a
+[klipper-lb](https://github.com/k3s-io/klipper-lb) pod to each nodes for each
+loadbalancer Service. The pod can insert iptables rules which DNAT ingress
+traffic to the nodeIP of the Service.
+
+Another option is manipulating the iptables rule by this controller itself.
+Since there is only one node, and the microshsift binary has the access to the
+host network.
+
+For those CNIs that relays on kube-proxy to support the kubernetes Service. We
+can also consider to use kube-proxy to do the job.
+
+### Update Status
+As, we don't want to ask user to allocated additional VIP as the loadbalancer
+IP. We can say that the ingress ip of the Service's `.status.loadBalancer` field
+shall always be the existing IPs of the node. When there are multiple IPs, the
+kubelet `NodeIP` shall be used by default.
+
+The controller shall also check if the Service port has already been allocated
+to other loadbalancer Service before updating the Service status.
+
+### Open Questions [optional]
+
+* Do we want to allow users to choose the IP other than the NodeIP as the
+Ingress IP of loadbalancer Service, when there are multiple IPs available?
+
```

Comment:
I would add an open question:

* How do we make this implementation compatible with the usage of metal-lb ?

Reply:
Do we need metal-lb if we have a load balancer built in?

Reply:
The function we're trying to implement is not the same as metal-lb. Metal-lb uses a VIP as the Ingress IP of loadbalancer Service, and advises it with BGP or ARP. So I am unsure if any customer wants to use metal-lb with microshift. So, maybe we can give a flag that users can disable the local loadbalancer controller. Users can choose if they want to use it or not.

Reply:
Something simple like being able to turn it off makes sense. Let's keep this simple for now, though, and not commit to doing that in the first version. When we have a definite requirement to support metal-lb, we can evaluate the best way to do it.